proto/api_v3/query_service.proto

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax="proto3";
+
+package jaeger.api_v3;
+
+import "opentelemetry/proto/trace/v1/trace.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+
+option go_package = "api_v3";
+option java_package = "io.jaegertracing.api_v3";
+
+// Request object to get a trace.
+message GetTraceRequest {
+  // Hex encoded 128 bit trace_id
+  string trace_id = 1;
+}
+
+// A single response chunk holds a single trace.
+// Note that a subsequent request to get a trace might return a different result
+// because spans are reported in asynchronous chunks.
+message SpansResponseChunk {
+  // OTLP resource spans
+  // see https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto
+  // In case of JSON format the ids (trace_id, span_id, parent_id) are encoded in base64 even though OpenTelemetry specification
+  // mandates to use hex encoding (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).
+  // Base64 is chosen to keep compatibility with JSONPb codec.
+  repeated opentelemetry.proto.trace.v1.ResourceSpans resource_spans = 1;
+}
+
+// Query parameters to find traces.
+message TraceQueryParameters {
+  string service_name = 1;
+  string operation_name = 2;
+  map<string, string> attributes = 3;
```

Comment:
Should we clarify which attributes these are supposed to match? We've been pretty loose about it in the original API, leaving the interpretation to the storage. I.e. should all these attributes match on a single span, or could they match across spans? Do they match span attributes only or span logs as well?

Reply:
The match should be done on tags and process tags. ES supports match on logs as well when kibana support is enabled (flat schema).

Reply:
In ES they must match on a single span. How is it done in cassandra? I think all storages should follow this.

Reply:
I don't think this is true in Cassandra, because it takes the tag=value string and looks up an index that just gives trace IDs. If more than one tag is provided, it could easily match on different spans.

Reply:
I have added a comment on the `TraceQueryParameters` that clarifies that some storage implementations might deviate.

Alright so C* does deviate as well. What was the original design for attributes? Match any attributes within trace or in a single span?

Reply:
Another way to see this: what would you expect as a user? If you specify two pairs of attributes, would you expect them to exist throughout the trace, or for all attributes to exist as part of the same span? I'm not quite sure I have an answer here... Here's one case advocating for attributes to exist throughout the trace:

- root span has "userID=123", no other spans contain this
- span (not root span) has error=true
- SRE is looking for traces for user 123 with errors in it

Reply:
If you want to create an issue to address this, this PR can be merged as is.